modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,217 @@
+
+import (
+	errorsmod "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	capabilitytypes "github.com/cosmos/ibc-go/modules/capability/types"
+	"github.com/cosmos/ibc-go/v7/modules/apps/callbacks/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v7/modules/core/05-port/types"
+	ibcerrors "github.com/cosmos/ibc-go/v7/modules/core/errors"
+	ibcexported "github.com/cosmos/ibc-go/v7/modules/core/exported"
+)
+
+var (
+	_ porttypes.Middleware            = (*IBCMiddleware)(nil)
+	_ porttypes.PacketDataUnmarshaler = (*IBCMiddleware)(nil)
+)
+
+// IBCMiddleware implements the ICS26 callbacks for the ibc-callbacks middleware given
+// the underlying application.
+type IBCMiddleware struct {
+	app     types.PacketUnmarshalerIBCModule
+	channel porttypes.ICS4Wrapper
+
+	contractKeeper types.ContractKeeper
+}
+
+// NewIBCMiddleware creates a new IBCMiddlware given the keeper and underlying application
+func NewIBCMiddleware(
+	app porttypes.IBCModule,
+	channel porttypes.ICS4Wrapper,
+	contractKeeper types.ContractKeeper,
+) IBCMiddleware {
+	packetUnmarshalerApp, ok := app.(types.PacketUnmarshalerIBCModule)
+	if !ok {
+		panic("underlying application does not implement PacketDataUnmarshaler")
+	}
+	return IBCMiddleware{
+		app:            packetUnmarshalerApp,
+		channel:        channel,
+		contractKeeper: contractKeeper,
+	}
+}
+
+// UnmarshalPacketData defers to the underlying app to unmarshal the packet data.
+func (im IBCMiddleware) UnmarshalPacketData(bz []byte) (interface{}, error) {
+	return im.app.UnmarshalPacketData(bz)
+}
+
+// OnAcknowledgementPacket implements wasm callbacks for acknowledgement packets.
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	// we first call the underlying app to handle the acknowledgement
+	appResult := im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if appResult != nil {
+		return appResult
+	}
+
+	callbackData, err := types.GetCallbackData(im.app, packet, ctx.GasMeter().GasRemaining())
+	if err != nil {
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData.ContractAddr, callbackData.GasLimit, err)
+		return appResult
+	}
+
+	var ack channeltypes.Acknowledgement
+	if err := channeltypes.SubModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		err = errorsmod.Wrapf(ibcerrors.ErrUnknownRequest, "cannot unmarshal callback packet acknowledgement: %v", err)
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData.ContractAddr, callbackData.GasLimit, err)
+		return appResult
+	}
+
+	err = im.contractKeeper.IBCAcknowledgementPacketCallback(ctx, packet, callbackData.CustomMsg, ack, relayer, callbackData.ContractAddr, callbackData.GasLimit)
```

Comment:
Why don't you just pass in the callbackData to the callbacks in the contractKeeper?

Doesn't make sense to pack in these fields from the packet into a custom type only to unpack them again when you pass to the keeper function